### PR TITLE
[opentitanlib] Adds session proxy communication protocol

### PR DESCRIPTION
**Pull request by semihalf-mazurek-michal**
Introduces a set of serializable struct/enum as implementation of
communication protocol used between 'Proxy Transport Instance' and
opentitansession process.

This protocol is part of issues: lowRISC#10889, lowRISC#10217

Signed-off-by: Michał Mazurek <maz@semihalf.com>
Signed-off-by: Jes B. Klinke <jbk@chromium.org>

### CODE REVIEW

**Review by jesultra — APPROVED** [review]
Looks good.

For demonstation on how this is going to be used, see:
https://github.com/jesultra/opentitan/commits/session

**Review by timothytrippel — APPROVED** [review]
LGTM, thanks @semihalf-mazurek-michal !

File: sw/host/opentitanlib/src/lib.rs [hunk 1]
```diff
@@ -11,3 +11,4 @@ pub mod otp;
 pub mod spiflash;
 pub mod transport;
 pub mod util;
+pub mod proxy;
```
[the remark this answers is not in the document]
Comment (timothytrippel) [reply sw/host/opentitanlib/src/lib.rs]: I think you can use `rustfmt <filename>` to take care of this? (@mcy can confirm?) This should ideally be run on all rust files.

[the remark this answers is not in the document]
Comment (semihalf-mazurek-michal) [reply sw/host/opentitanlib/src/lib.rs]: I usually use cargo fmt - but in this case it generate a lot of unrelated changes through the entire library, so i only committed changes for this file. 
After executing the clippy command I think there are many places in opentitanlib where there is room for improvement.
But i don't think this commit should do everything everywhere.

**Review by mcy — APPROVED** [review]
Hit this with rustfmt?

**Comment by semihalf-mazurek-michal** [comment]
I 'hit' my commits with rustfmt bit it generate a loot of changes:

`       modified:   src/app/mod.rs
	modified:   src/backend/cw310.rs
	modified:   src/backend/hyperdebug.rs
	modified:   src/backend/ultradebug.rs
	modified:   src/backend/verilator.rs
	modified:   src/io/uart.rs
	modified:   src/transport/common/uart.rs
	modified:   src/transport/cw310/mod.rs
	modified:   src/transport/hyperdebug/i2c.rs
	modified:   src/transport/hyperdebug/mod.rs
	modified:   src/transport/ultradebug/gpio.rs
	modified:   src/transport/ultradebug/mod.rs
	modified:   src/transport/ultradebug/spi.rs
	modified:   src/transport/ultradebug/uart.rs
	modified:   src/transport/verilator/uart.rs
	modified:   src/util/usb.rs`
	
In this case it will be better to make it in separate PR with only contain fix for rust formatting.

File: sw/host/opentitanlib/src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (jesultra) [reply sw/host/opentitanlib/src/lib.rs]: Right, rustfmt for just the one file protocol.rs is fine.  We do not expect this PR to clean up every other file.  (And my personal opinion is that you should not in general make commits that only change formatting.  They introduce too much clutter in the revision history.  Instead, fix formatting only when touching the file/section anyway.)

**Comment by semihalf-mazurek-michal** [comment]
@mcy @timothytrippel 
Following PR https://github.com/lowRISC/opentitan/pull/11056 contain fix of rust formatting.

**Pull request merged**